README
Liblazy - D-Bus methods provided for convenience
================================================

Liblazy is a simple and easy to use library that provides convenient
functions for sending messages over the D-Bus daemon, querying information
from HAL or asking PolicyKit for a privilege. Its features may grow as
needed, though.


Rationale
=========

Once in a while, I stumble across an application which has no proper D-Bus
support and gathers some information through reading /proc or sysfs, or
any other circumstantial way, although software like HAL or PolicyKit
already have this information. Or they just want to send a simple message
to another application, sometimes even without caring about the reply.
These applications often only have to do very simple tasks.

So to adopt these applications, you always have to duplicate lots of code
over and over again.

Just some examples...

To to send a message over D-Bus one has to do:
  1. Get a D-Bus connection
  2. Check for errors
  3. Send a message over the bus
  4. Check for errors
  5. Get the reply
  6. Check for errors

To get a simple property from HAL, one has to do (with libhal in
this example):
  1. Get a D-Bus connection
  2. Check for errors
  3. Get a LibHalContext
  4. C.f.e.
  5. Register the connection with the context
  6. C.f.e.
  7. Fetch the property with libhal
  8. C.f.e

When you have to check if your application possess a specific privilege
according to PolicyKit:
    Either do a mixture of the above things to ask directly via D-Bus or
    link against libpolkit and do a complicated call of
    libpolkit_is_uid_allowed_for_privilege(7 arguments follow). This call
    gives you a lot of information you might not care about at all.</li>

So in every application, you have to implement functions like:</p>
   - dbus_send_message(service, path, interface, method);
   - dbus_send_message_with_reply(service, path, interface, method, &reply);
   - hal_get_property_int(udi, property)
   - ...

There are a lot of applications which just don't want to care too much
about D-Bus or HAL at all. I'm mostly talking about small applications, of
course, even about those without a mainloop. They just want to make use of
hardware databases and abstractions and a new way of inter process
communication we have with D-Bus these days.

So what I've done is combining these very simple functions into one small
library called liblazy. A small application may just want to do the
following to send a message over the bus and get its reply, maybe even
with ignoring any errors:

  DBusMessage *reply;
  liblazy_dbus_system_send_method_call("org.freedesktop.NetworkworkManager",
                                "/org/freedesktop/NetworkManager",
                                "org.freedesktop.NetworkManager",
                                "sleep",
                                &reply,
                                DBUS_TYPE_INVALID);

  int *arg_from_reply;
  liblazy_dbus_message_get_arg(reply, DBUS_TYPE_INT32, &arg_from_reply, 0);

That's it. The application developer just has to link against libdbus and
liblazy, without caring about any other stuff.

Liblazy also contains very simple functions to get a property from HAL or
ask for a privilege from PolicyKit. But it doesn't contain complex
functionality. It may grow as needed, though. But...

It is definitely not intended for replacing any of the dbus bindings,
libhal or libpolkit. Complex applications can and will still have their
own implementations and will link agains libhal or libpolkit or
whatever. They are free to use liblazy, of course. But it is definitely of
benefit for small applications which only need a subset of functionality
and like to profit from new software like D-Bus, HAL, etc. but without
having to care a lot. It should serve as a set of functions for developers
provided for convenience. Also a lot more command line applications could
make use of HAL, etc. in an easy way IMO. Additionally, it may serve as a
code example for developers to get familiar with D-Bus.


NEWS
- converted to PokicyKit >= 0.4 API
- add method to use private dbus connection


tools/test.c
#include "liblazy.h"

#include <stdio.h>
#include <unistd.h>

#define DBUS_NM_SERVICE			"org.freedesktop.NetworkManager"
#define DBUS_NM_INTERFACE		"org.freedesktop.NetworkManager"
#define DBUS_NM_PATH			"/org/freedesktop/NetworkManager"
#define DBUS_HAL_MANAGER_INTERFACE	"org.freedesktop.Hal.Manager"
#define DBUS_HAL_DEVICE_INTERFACE	"org.freedesktop.Hal.Device"
#define DBUS_HAL_UDI_COMPUTER		"/org/freedesktop/Hal/devices/computer"
#define DBUS_HAL_INTERFACE		"org.freedesktop.Hal"
#define DBUS_HAL_CPUFREQ_INTERFACE	"org.freedesktop.Hal.Device.CPUFreq"
#define DBUS_SYSTEM_PM_INTERFACE	"org.freedesktop.Policy.Power"

int main(int argc, char *argv[])
{
	char *governor = "ondemand";
	char *rpstring = NULL;
	int intval = -1;

	int ret = 0;
	DBusMessage *reply = NULL;

	printf("Try to set the ondemand governor\n");
/*	liblazy_dbus_system_use_private_connection(1);*/
	ret = liblazy_dbus_system_send_method_call(DBUS_HAL_INTERFACE,
						   DBUS_HAL_UDI_COMPUTER,
						   DBUS_HAL_CPUFREQ_INTERFACE,
						   "SetCPUFreqGovernor",
						   &reply,
						   DBUS_TYPE_STRING,
						   &governor,
						   DBUS_TYPE_INVALID);
	printf("return value: %d\n", ret);
	if (!ret)
		dbus_message_unref(reply);

	printf("------------------\n");
	printf("Try to get the current governor\n");
	ret = liblazy_dbus_system_send_method_call(DBUS_HAL_INTERFACE,
						   DBUS_HAL_UDI_COMPUTER,
						   DBUS_HAL_CPUFREQ_INTERFACE,
						   "GetCPUFreqGovernor",
						   &reply,
						   DBUS_TYPE_INVALID);

	ret = liblazy_dbus_message_get_basic_arg(reply, DBUS_TYPE_STRING, &rpstring, 0);
	printf("returned string: %s\n", rpstring);
	printf("return value: %d\n", ret);
	if (!ret)
		dbus_message_unref(reply);

	printf("--------------------\n");
	ret = liblazy_hal_get_property_string("/org/freedesktop/Hal/devices/computer",
					      "system.kernel.name",
					      &rpstring);
	printf("return value: %d\n", ret);
	printf("returned string: %s\n", rpstring);

	printf("string: %s\n", rpstring);
	liblazy_free_string(rpstring);

	printf("--------------------\n");
	printf("Try to get a string property from HAL\n");
	ret = liblazy_hal_get_property_string("/org/freedesktop/Hal/devices/computer",
					      "system.kernel.name",
					      &rpstring);
	printf("return value: %d\n", ret);
	printf("returned string: %s\n", rpstring);
	liblazy_free_string(rpstring);

	printf("--------------------\n");
	printf("Try to get a boolean value from HAL\n");
	ret = liblazy_hal_get_property_bool("/org/freedesktop/Hal/devices/computer",
					      "power_management.can_hibernate",
					      &intval);
	printf("return value: %d\n", ret);
	printf("returned int: %d\n", intval);

	printf("--------------------\n");
	printf("Try to get a integer value from HAL\n");
	ret = liblazy_hal_get_property_int("/org/freedesktop/Hal/devices/acpi_BAT0",
					    "battery.charge_level.current",
					    &intval);
	printf("return value: %d\n", ret);
	printf("returned int: %d\n", intval);


	printf("--------------------\n");
	ret = liblazy_hal_get_property_string("/org/freedesktop/Hal/devices/computer",
					      "system.kernel.name",
					      &rpstring);
	printf("return value: %d\n", ret);
	printf("returned string: %s\n", rpstring);

	printf("string: %s\n", rpstring);
	liblazy_free_string(rpstring);

	printf("--------------------\n");
	printf("Try to get a string property from HAL\n");
	ret = liblazy_hal_get_property_string("/org/freedesktop/Hal/devices/computer",
					      "system.kernel.name",
					      &rpstring);
	printf("return value: %d\n", ret);
	printf("returned string: %s\n", rpstring);
	liblazy_free_string(rpstring);

	printf("--------------------\n");
	printf("Try to get a boolean value from HAL\n");
	ret = liblazy_hal_get_property_bool("/org/freedesktop/Hal/devices/computer",
					      "power_management.can_hibernate",
					      &intval);
	printf("return value: %d\n", ret);
	printf("returned int: %d\n", intval);

	printf("--------------------\n");
	printf("Trying to send signal over system bus\n");
	ret = liblazy_dbus_system_send_signal("/macht/holger/super",
					      "org.holger.super",
					      "heyho",
					      DBUS_TYPE_INVALID);
	printf("return value: %d\n", ret);
	

	return 0;
}


liblazy/liblazy.c
/***************************************************************************
 *                                                                         *
 *                              liblazy                                    *
 *                                                                         *
 *           Copyright (C) 2006 Holger Macht <holger@homac.de>             *
 *                                                                         *
 *              Author(s): Holger Macht <holger@homac.de>                  *
 *                                                                         *
 * This library is free software; you can redistribute it and/or modify it *
 * under the terms of the GNU Lesser General Public License as published   *
 * by the Free Software Foundation; either version 2.1 of the License, or  *
 * (at your option) any later version.                                     *
 *                                                                         *
 * This library is distributed in the hope that it will be useful, but     *
 * WITHOUT ANY WARRANTY; without even the implied warranty of              *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU       *
 * Lesser General Public License for more details.                         *
 *                                                                         *
 * You should have received a copy of the GNU Lesser General Public        *
 * License along with this library; if not, write to the Free Software     *
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA           *
 * 02110-1301  USA                                                         *
 *                                                                         *
 ***************************************************************************/

#include "liblazy.h"

#include <stdlib.h>

void liblazy_free_string(char *string)
{
	if (string == NULL)
		return;
	free(string);
	string = NULL;
}

void liblazy_free_strlist(char **strlist)
{
	int i = 0;
	if (strlist == NULL)
		return;
	for (i = 0; strlist[i] != NULL; i++) {
		free(strlist[i]);
		strlist[i] = NULL;
	}
	free(strlist);
	strlist = NULL;
}

/** @mainpage liblazy
 *
 * \section sec_intro Introduction
 *
 * D-Bus methods provided for convenience
 * Liblazy is a simple and easy to use library that provides convenient
 * functions for sending messages over the D-Bus daemon, querying
 * information from HAL or asking PolicyKit for a privilege.
 */


liblazy/liblazy.h
/***************************************************************************
 *                                                                         *
 *                              liblazy                                    *
 *                                                                         *
 *           Copyright (C) 2006 Holger Macht <holger@homac.de>             *
 *                                                                         *
 *              Author(s): Holger Macht <holger@homac.de>                  *
 *                                                                         *
 * This library is free software; you can redistribute it and/or modify it *
 * under the terms of the GNU Lesser General Public License as published   *
 * by the Free Software Foundation; either version 2.1 of the License, or  *
 * (at your option) any later version.                                     *
 *                                                                         *
 * This library is distributed in the hope that it will be useful, but     *
 * WITHOUT ANY WARRANTY; without even the implied warranty of              *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU       *
 * Lesser General Public License for more details.                         *
 *                                                                         *
 * You should have received a copy of the GNU Lesser General Public        *
 * License along with this library; if not, write to the Free Software     *
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA           *
 * 02110-1301  USA                                                         *
 *                                                                         *
 ***************************************************************************/

#ifndef LIBLAZY_H
#define LIBLAZY_H

#define DBUS_API_SUBJECT_TO_CHANGE 1
#include <dbus/dbus.h>

#ifdef __cplusplus
extern "C" {
#endif /* __cplusplus */

/** @defgroup liblazy Liblazy - D-Bus methods provided for convenience
 * @{ */

#define LIBLAZY_ERROR_GENERAL			-1
#define LIBLAZY_ERROR_INVALID_ARGUMENT		-2
#define LIBLAZY_ERROR_HAL_NOT_READY		-10
#define LIBLAZY_ERROR_HAL_NO_SUCH_PROPERTY	-11

#define LIBLAZY_ERROR_DBUS_NOT_READY		-20
#define LIBLAZY_ERROR_DBUS_NO_REPLY		-21
#define LIBLAZY_ERROR_DBUS_ERROR_IS_SET		-22

/** @brief free a string
 *
 * @param string the string to free
 */
void liblazy_free_string(char *string);

/** @brief free a null terminated array of strings
 *
 * @param strlist the string list to free
 */
void liblazy_free_strlist(char **strlist);

/** @brief send a method call to the system bus
 *
 * sends a method call to the system bus. The call blocks if a reply is
 * given. If the call shouldn't block, give NULL for the reply
 *
 * @param destination the destination to send to
 * @param path the object path to send to
 * @param interface the interface to send to
 * @param method the method to send
 * @param reply a DBusMessage to store the reply or NULL if the call
 *              shouldn't block
 * @param first_arg_type a DBUS_TYPE_* of the fist argument
 * @param ... variable argument list finished with DBUS_TYPE_INVALID
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_dbus_system_send_method_call(const char *destination, const char *path,
					 const char *interface, const char *method,
					 DBusMessage **reply,
					 int first_arg_type, ...);

/** @brief send a method call to the session bus
 *
 * sends a method call to the session bus. The call blocks if a reply is
 * given. If the call shouldn't block, give NULL for the reply
 *
 * @param destination the destination to send to
 * @param path the object path to send to
 * @param interface the interface to send to
 * @param method the method to send
 * @param reply a DBusMessage to store the reply or NULL if the call
 *              shouldn't block
 * @param first_arg_type a DBUS_TYPE_* of the fist argument
 * @param ... variable argument list finished with DBUS_TYPE_INVALID
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */

int liblazy_dbus_session_send_method_call(const char *destination, const char *path,
					  const char *interface, const char *method,
					  DBusMessage **reply,
					  int first_arg_type, ...);

/** @brief send a signal over the system bus
 *
 * sends a signal over the system bus. The call blocks if a reply is
 * given. If the call shouldn't block, give NULL for the reply
 *
 * @param path the object path to send to
 * @param interface the interface to send to
 * @param name the name of the signal
 * @param first_arg_type a DBUS_TYPE_* of the fist argument
 * @param ... variable argument list finished with DBUS_TYPE_INVALID
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_dbus_system_send_signal(const char *path, const char *interface,
				    const char *name, int first_arg_type, ...);

/** @brief send a signal over the session bus
 *
 * sends a signal over the session bus. The call blocks if a reply is
 * given. If the call shouldn't block, give NULL for the reply
 *
 * @param path the object path to send to
 * @param interface the interface to send to
 * @param name the name of the signal
 * @param first_arg_type a DBUS_TYPE_* of the fist argument
 * @param ... variable argument list finished with DBUS_TYPE_INVALID
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_dbus_session_send_signal(const char *path, const char *interface,
				     const char *name, int first_arg_type, ...);

/** @brief get an argument from a DBusMessage
 *
 * @param message the DBusMessage to get the argument from
 * @param type the type of the argument
 * @param arg a location to store the argument
 * @param no a number specifying the n'th argument of the given type
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_dbus_message_get_basic_arg(DBusMessage *message, int type,
				       void *arg, int no);

/** @brief get a array argument from a DBusMessage as a string list
 *
 * @param message the message to get the argument from
 * @param strlist a pointer to a string array to store the result
 * @param strlist pointer to array of strings to store the result. Has to
 *		  be freed with @ref liblazy_free_strlist
 * @param no a number specifying the n'th string array in the reply
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_dbus_message_get_strlist_arg(DBusMessage *message,
					 char ***strlist, int no);

/** @brief use a private connection for system bus messages
 *
 * Call this function with a boolean value to tell the library whether to
 * use a private connection for system bus messages
 * (liblazy_dbus_system*). Defaults to false. If a private connection is
 * used, the libarary creates a new connection on _every_ call and
 * immediately closes it. So this should be used very carefully. Only use
 * if your application does very rare dbus calls and doesn't have a
 * mainloop to listen if D-Bus died but still should be able to survive
 * D-Bus restarts.
 *
 * @param use_privat 1 if the library should use a private connection, 0
 *		     otherwise
 */
void liblazy_dbus_system_use_private_connection(int use_private);

/** @brief get integer property from HAL
 *
 * fetches one interger value from HAL.
 *
 * @param udi the device to fetch the value from
 * @param property the property to fetch
 * @param value location to store the result
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_hal_get_property_int(char *udi, char *property, int *value);

/** @brief get boolean property from HAL
 *
 * fetches one boolean value from HAL.
 *
 * @param udi the device to fetch the value from
 * @param property the property to fetch
 * @param value location to store the result
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_hal_get_property_bool(char *udi, char *property, int *value);

/** @brief get string property from HAL
 *
 * fetches one string value from HAL.
 *
 * @param udi the device to fetch the value from
 * @param property the property to fetch
 * @param value location to store the result
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_hal_get_property_string(char *udi, char *property, char **value);

/** @brief get string list property from HAL
 *
 * fetches a string list value from HAL.
 *
 * @param udi the device to fetch the value from
 * @param property the property to fetch
 * @param strlist pointer to array of strings to store the result. Has to
 *		  be freed with @ref liblazy_free_strlist
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_hal_get_property_strlist(char *udi, char *property, char ***strlist);

/** @brief check if a device has a capability
 *
 * @param udi the device to query on
 * @param capability the capability to query for
 *
 * @return 1 if the device possesses the capability, 0 if not and
 *         LIBLAZY_ERROR_* on failure
 */
int liblazy_hal_query_capability(char *udi, char *capability);

/** @brief find devices with a given capability
 *
 * @param capability the capability the devices should have
 * @param strlist pointer to array of strings to store the result. Has to
 *		  be freed with @ref liblazy_free_strlist
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_hal_find_device_by_capability(char *capability, char ***strlist);

/** @brief find devices with given key and value
 *
 * @param key the key to match against
 * @param value the value to match against
 * @param strlist pointer to array of strings to store the result. Has to
 *		  be freed with @ref liblazy_free_strlist
 *
 * @return 0 on success, LIBLAZY_ERROR_* on failure
 */
int liblazy_hal_find_device_by_string_match(char *key, char *value, char ***strlist);

/** @brief check if a user possesses a privilege
 *
 * Check if the caller possesses the given privilege on the default device
 * '/org/freedesktop/Hal/devices/computer'
 *
 * @param privilege the privilege to check for
 *
 * @return 1 if the caller is privileged, 0 if not, and LIBLAZY_ERROR_* on failure
 */
int liblazy_hal_is_caller_privileged(char *privilege);

#ifdef __cplusplus
}
#endif /* __cplusplus */

#endif				/* LIBLAZY_H */

/** @} */// end of liblazy group


liblazy/liblazy_local.h
/***************************************************************************
 *                                                                         *
 *                              liblazy                                    *
 *                                                                         *
 *           Copyright (C) 2006 Holger Macht <holger@homac.de>             *
 *                                                                         *
 *              Author(s): Holger Macht <holger@homac.de>                  *
 *                                                                         *
 * This library is free software; you can redistribute it and/or modify it *
 * under the terms of the GNU Lesser General Public License as published   *
 * by the Free Software Foundation; either version 2.1 of the License, or  *
 * (at your option) any later version.                                     *
 *                                                                         *
 * This library is distributed in the hope that it will be useful, but     *
 * WITHOUT ANY WARRANTY; without even the implied warranty of              *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU       *
 * Lesser General Public License for more details.                         *
 *                                                                         *
 * You should have received a copy of the GNU Lesser General Public        *
 * License along with this library; if not, write to the Free Software     *
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA           *
 * 02110-1301  USA                                                         *
 *                                                                         *
 ***************************************************************************/

#ifndef LIBLAZY_LOCAL_H
#define LIBLAZY_LOCAL_H

#define ERROR(string, args...) \
do{\
	fprintf(stderr, "liblazy (%s:%d): "string"\n", __FUNCTION__, __LINE__, ## args); \
}while(0);

int liblazy_dbus_send_method_call(const char *destination, const char *path,
				  const char *interface, const char *method,
				  int bus_type,
				  DBusMessage **reply,
				  int first_arg_type, va_list var_args);

#endif /* LIBLAZY_LOCAL_H */


liblazy/liblazy_hal.c
/***************************************************************************
 *                                                                         *
 *                              liblazy                                    *
 *                                                                         *
 *         Copyright (C) 2006,2007 Holger Macht <holger@homac.de>          *
 *                                                                         *
 *              Author(s): Holger Macht <holger@homac.de>                  *
 *                                                                         *
 * This library is free software; you can redistribute it and/or modify it *
 * under the terms of the GNU Lesser General Public License as published   *
 * by the Free Software Foundation; either version 2.1 of the License, or  *
 * (at your option) any later version.                                     *
 *                                                                         *
 * This library is distributed in the hope that it will be useful, but     *
 * WITHOUT ANY WARRANTY; without even the implied warranty of              *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU       *
 * Lesser General Public License for more details.                         *
 *                                                                         *
 * You should have received a copy of the GNU Lesser General Public        *
 * License along with this library; if not, write to the Free Software     *
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA           *
 * 02110-1301  USA                                                         *
 *                                                                         *
 ***************************************************************************/

#include "liblazy.h"
#include "liblazy_local.h"

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#define DBUS_HAL_SERVICE		"org.freedesktop.Hal"
#define DBUS_HAL_DEVICE_INTERFACE	"org.freedesktop.Hal.Device"
#define DBUS_HAL_MANAGER_PATH		"/org/freedesktop/Hal/Manager"
#define DBUS_HAL_MANAGER_INTERFACE	"org.freedesktop.Hal.Manager"
#define DBUS_HAL_COMPUTER_PATH		"/org/freedesktop/Hal/devices/computer"

static int liblazy_hal_property_exists(char *udi, char *property)
{
	int		error	= 0;
	int		exists	= 0;
	DBusMessage	*reply;

	if (udi == NULL || property == NULL)
		return LIBLAZY_ERROR_INVALID_ARGUMENT;

	error = liblazy_dbus_system_send_method_call(DBUS_HAL_SERVICE,
						     udi,
						     DBUS_HAL_DEVICE_INTERFACE,
						     "PropertyExists",
						     &reply,
						     DBUS_TYPE_STRING,
						     &property,
						     DBUS_TYPE_INVALID);

	if (error) {
		ERROR("Error checking if property '%s' exists", property);
		return error;
	}

	error = liblazy_dbus_message_get_basic_arg(reply, DBUS_TYPE_BOOLEAN,
						   &exists, 0);
	if (reply != NULL)
		dbus_message_unref(reply);
	if (exists == 1)
		return 1;
	return error;
}

static int liblazy_hal_get_property(char *udi, char *property, char *method,
				    int type, void *value)
{
	int		error = 0;
	DBusMessage	*reply;

	if (udi == NULL || property == NULL )
		return LIBLAZY_ERROR_INVALID_ARGUMENT;

	error = liblazy_dbus_system_send_method_call(DBUS_HAL_SERVICE,
						     udi,
						     DBUS_HAL_DEVICE_INTERFACE,
						     method,
						     &reply,
						     DBUS_TYPE_STRING,
						     &property,
						     DBUS_TYPE_INVALID);	

	if (error) {
		ERROR("Error sending '%s' to HAL", method);
		return error;
	}

	error = liblazy_dbus_message_get_basic_arg(reply, type, value, 0);
	if (reply != NULL)
		dbus_message_unref(reply);
	if (error)
		ERROR("Error fetching property '%s'", property);
	return error;
}

static int liblazy_hal_get_strlist_manager(char ***strlist, char *method,
					   int first_arg_type, ...)
{
	int		error	= 0;
	DBusMessage	*reply;
	va_list		var_args;

	va_start(var_args, first_arg_type);

	error = liblazy_dbus_send_method_call(DBUS_HAL_SERVICE,
					      DBUS_HAL_MANAGER_PATH,
					      DBUS_HAL_MANAGER_INTERFACE,
					      method,
					      DBUS_BUS_SYSTEM,
					      &reply,
					      first_arg_type,
					      var_args);
	va_end(var_args);

	if (error) {
		ERROR("Error while sending method %s to HAL", method);
		return error;
	}

	error = liblazy_dbus_message_get_strlist_arg(reply, strlist, 0);
	if (reply != NULL)
		dbus_message_unref(reply);
	return error;
}

int liblazy_hal_get_property_string(char *udi, char *property, char **value)
{
	char	*str;
	int	ret;

	ret = liblazy_hal_property_exists(udi, property);
	if (ret != 1) {
		if (ret == 0)
			ret = LIBLAZY_ERROR_HAL_NO_SUCH_PROPERTY;
		goto Error;
	}

	ret = liblazy_hal_get_property(udi, property, "GetPropertyString",
				       DBUS_TYPE_STRING, &str);
	if (ret)
		goto Error;

	*value = strdup(str);
	return ret;
Error:
	*value = NULL;
	return ret;
}

int liblazy_hal_get_property_int(char *udi, char *property, int *value)
{
	int ret;

	ret = liblazy_hal_property_exists(udi, property);
	if (ret != 1) {
		if (ret == 0)
			ret = LIBLAZY_ERROR_HAL_NO_SUCH_PROPERTY;
		*value = -1;
		return ret;
	}

	return liblazy_hal_get_property(udi, property, "GetPropertyInteger",
					DBUS_TYPE_INT32, value);
}

int liblazy_hal_get_property_bool(char *udi, char *property, int *value)
{
	int ret;

	ret = liblazy_hal_property_exists(udi, property);
	if (ret != 1) {
		if (ret == 0)
			ret = LIBLAZY_ERROR_HAL_NO_SUCH_PROPERTY;
		*value = -1;
		return ret;
	}

	return liblazy_hal_get_property(udi, property, "GetPropertyBoolean",
					DBUS_TYPE_BOOLEAN, value);
}

int liblazy_hal_get_property_strlist(char *udi, char *property, char ***strlist)
{
	int		error	= 0;
	DBusMessage	*reply;

	if (udi == NULL || property == NULL )
		return LIBLAZY_ERROR_INVALID_ARGUMENT;

	error = liblazy_hal_property_exists(udi, property);
	if (error != 1) {
		if (error == 0)
			error = LIBLAZY_ERROR_HAL_NO_SUCH_PROPERTY;
		goto Error;
	}

	error = liblazy_dbus_system_send_method_call(DBUS_HAL_SERVICE,
						     udi,
						     DBUS_HAL_DEVICE_INTERFACE,
						     "GetPropertyStringList",
						     &reply,
						     DBUS_TYPE_STRING,
						     &property,
						     DBUS_TYPE_INVALID);	

	if (error) {
		ERROR("Error while getting strlist property %s from HAL", property);
		goto Error;
	}

	error = liblazy_dbus_message_get_strlist_arg(reply, strlist, 0);
	if (reply != NULL)
		dbus_message_unref(reply);
	if (error)
		goto Error;
	return error;
 
Error:
	strlist[0] = NULL;
	return error;
}

int liblazy_hal_query_capability(char *udi, char *capability)
{
	int	i;
	int	error	= -1;
	char	**caps	= NULL;

	if (udi == NULL || capability == NULL )
		return LIBLAZY_ERROR_INVALID_ARGUMENT;

	error = liblazy_hal_get_property_strlist(udi, "info.capabilities", &caps);
	if (error)
		return error;

	if (caps != NULL) {
		for (i = 0; caps[i] != NULL; i++) {
			if (strcmp(caps[i], capability) == 0)
				error = 1;
		}
		liblazy_free_strlist(caps);
	}

	return error;
}

int liblazy_hal_find_device_by_capability(char *capability, char ***strlist)
{
	int error;
	error = liblazy_hal_get_strlist_manager(strlist, "FindDeviceByCapability",
						DBUS_TYPE_STRING, &capability,
						DBUS_TYPE_INVALID);
	if (error) {
		strlist[0] = NULL;
		*strlist = NULL;
	}
	return error;

}

int liblazy_hal_find_device_by_string_match(char *key, char *value, char ***strlist)
{
	int error;
	error = liblazy_hal_get_strlist_manager(strlist, "FindDeviceStringMatch",
						DBUS_TYPE_STRING, &key,
						DBUS_TYPE_STRING, &value, DBUS_TYPE_INVALID);
	if (error) {
		strlist[0] = NULL;
		*strlist = NULL;
	}
	return error;
}

int liblazy_hal_is_caller_privileged(char *privilege)
{
	DBusMessage	*reply;
	DBusError	dbus_error;
	DBusConnection	*dbus_connection;
	const char	*unique_name;
	char		*allowed;
	int		error = 0;

	if (privilege == NULL )
		return LIBLAZY_ERROR_INVALID_ARGUMENT;

	dbus_error_init(&dbus_error);

	dbus_connection = dbus_bus_get(DBUS_BUS_SYSTEM, &dbus_error);
	if (dbus_error_is_set(&dbus_error)) {
		ERROR("Connection to dbus not ready, skipping privilege "
		      "lookup for privilege %s: %s\n",
		      privilege, dbus_error.message);
		dbus_error_free(&dbus_error);
		return LIBLAZY_ERROR_DBUS_NOT_READY;
	}

	unique_name = dbus_bus_get_unique_name(dbus_connection);

	error = liblazy_dbus_system_send_method_call(DBUS_HAL_SERVICE,
						     DBUS_HAL_COMPUTER_PATH,
						     DBUS_HAL_DEVICE_INTERFACE,
						     "IsCallerPrivileged",
						     &reply,
						     DBUS_TYPE_STRING, &privilege, 
						     DBUS_TYPE_STRING, &unique_name, 
						     DBUS_TYPE_INVALID);

	if (error)
		return error;

	error = liblazy_dbus_message_get_basic_arg(reply, DBUS_TYPE_STRING,
						   &allowed, 0);

	if (reply != NULL)
		dbus_message_unref(reply);
	if (error)
		return error;

	if (strcmp(allowed, "yes") == 0)
		return 1;
	return 0;
}
